### PR TITLE
#458 - allow user to set formenctype

### LINKED ISSUE #458: Support multipart/form-data as encoding type on form component
What are you building with SQLPage ?

A utility to merge yaml files.

In this use case, copy-paste is unfortunately the most straightforward way to interact with the UI. Uploading the files to merge is too hard.

The files are fairly small, but are large enough to hit the URL encoded data limit.

Using multipart/form-data as encoding type would be ideal in this case.

For now I've patched the form component:
```diff
--- sqlpage/templates/form.handlebars	2024-06-25 10:14:48.295321207 +1200
+++ /home/daniels/git/SQLpage/sqlpage/templates/form.handlebars	2024-06-24 09:31:31.023355037 +1200
@@ -2,7 +2,6 @@
     {{#if id}}id="{{id}}"{{/if}}
     class="my-3 {{class}}"
     method="{{default method "post"}}"
-    {{#if formenctype}}formenctype="{{formenctype}}"{{/if}}
     {{#if action}}action="{{action}}"
     {{else}}
     {{#if id}}action="#{{id}}"{{/if}}
@@ -119,7 +118,7 @@
                 </label>
             {{/if}}
             {{/if}}
-            {{#if and (eq type "file") (not formenctype)}}
+            {{#if (eq type "file")}}
                 <!-- Change the form encoding type if this is a file input-->
                 {{#delay}}formenctype="multipart/form-data"{{/delay}}
             {{/if}}
@@ -132,7 +131,6 @@
             {{#if validate_outline}} btn-outline-{{validate_outline}} {{/if}}
             {{#if validate_size}} btn-{{validate_size}} {{/if}}"
             {{flush_delayed}}
-            {{#if formenctype}}formenctype="{{formenctype}}"{{/if}}
             type="submit"
             {{#if validate}}value="{{validate}}"{{/if}}>
         {{/if}}
@@ -146,4 +144,4 @@
             value="{{reset}}">
         {{/if}}
     </fieldset>
-</form>
\ No newline at end of file
+</form>
```
My patch is simple and just makes explicit encoding type take precedence.

**Describe the solution you'd like**

I'd like to be able to explicitly set the formenctype on the form component.

Currently this is only set when one of the form fields is ``file`` type.

I think this behaviour is good, but we should also allow the user to specify the formenctype also.

**Describe alternatives you've considered**

Uploading the YAML files instead.
Unfortunately this is much more steps for the user than simply copy-pasting the content from the text editor straight into the UI.

### PR DESCRIPTION
**Pull request by djyotta**
Closes #458

### CODE REVIEW

**Review by lovasoa — APPROVED** [review]
Good. Can we document this behavior somewhere? Maybe in one of the examples of the form component?

**Comment by djyotta** [comment]
@lovasoa will do. Still testing it. I had some trouble with applying it to the textarea input directly. Seems adding it to the submit button works.

I want to make sure this works with also a file input present as they would both trigger adding the attribute to the submit button...

**Comment by lovasoa** [comment]
Okay, good! Switching this pr to draft mode, then.
Feel free to switch it back whenever you want.

**Comment by djyotta** [comment]
@lovasoa after reading the following:
* https://www.w3schools.com/tags/att_input_formenctype.asp
* https://www.w3schools.com/tags/att_form_enctype.asp

I decided to expose the `enctype` as top-level attribute as `formenctype` is intended for `submit` and `image` type inputs only.

Further, I would suggest we remove the `formenctype` being added automatically when file type input is present. Even though it's nice for newbies, there are the following short-comings:
* it overrides the form `enctype`  - could be confusing when dev explicitly sets this
* it won't be automatically applied to other submit buttons (if any)

For now I updated the docs recommending to set it explicitly for file upload rather than rely on current behaviour. Then we can decide whether to alter existing behaviour independently of this PR

**Comment by lovasoa** [comment]
I understand what you mean. But the whole promise of SQLPage is to make it easy to build UIs on top of databases without any other knowledge than just sql. I really think file upload should work by default, and not require even knowing what formenctype means.

File: examples/official-site/sqlpage/migrations/01_documentation.sql [hunk 1]
```diff
@@ -211,7 +211,23 @@ INSERT INTO component(name, icon, description) VALUES
     'The value entered by the user in a field named x will be accessible to the target SQL page as a variable named $x.
     For instance, you can create a SQL page named "create_user.sql" that would contain "INSERT INTO users(name) VALUES($name)"
     and a form with its action property set to "create_user.sql" that would contain a field named "name".');
+INSERT INTO parameter(component, name, description_md, type, top_level, optional) SELECT 'form', * FROM (VALUES
+    -- top level
+    ('enctype', '
+When ``method="post"``, this specifies how the form-data should be encoded
+when submitting it to the server.
```
Comment (lovasoa) [inline examples/official-site/sqlpage/migrations/01_documentation.sql:218]: I think just this line is enough. Advanced users will find it when they need it, and we wont confuse others with things they don't need.

**Review by lovasoa — CHANGES_REQUESTED** [review]

File: examples/official-site/sqlpage/migrations/01_documentation.sql [hunk 2]
```diff
@@ -401,6 +417,15 @@ But note that SQLPage cookies already have the `SameSite=strict` attribute by de
 ## File upload
 
 You can use the `file` type to allow the user to upload a file.
+
+> NOTE: It is recommended to use ``multipart/form-data`` to encode the from
+data for file uploads. This is because the default encoding
+(``application/x-www-form-urlencoded``) applied to arbitrary binary can be up
+to 3 times the size of the file.
+For now, ``formenctype="multipart/form-data"`` is automatically applied to the
+default validate button - but this may change in the future.
```
[the remark this answers is not in the document]
Comment (djyotta) [reply examples/official-site/sqlpage/migrations/01_documentation.sql:426]: What about setting enctype on the form as multipart/form-data by default?

This line would still be technically true in that case but I will remove it anyway for ease of reading.

File: examples/official-site/sqlpage/migrations/01_documentation.sql [hunk 3]
```diff
@@ -401,6 +417,15 @@ But note that SQLPage cookies already have the `SameSite=strict` attribute by de
 ## File upload
 
 You can use the `file` type to allow the user to upload a file.
+
+> NOTE: It is recommended to use ``multipart/form-data`` to encode the from
+data for file uploads. This is because the default encoding
+(``application/x-www-form-urlencoded``) applied to arbitrary binary can be up
+to 3 times the size of the file.
```
[the remark this answers is not in the document]
Comment (djyotta) [reply examples/official-site/sqlpage/migrations/01_documentation.sql:424]: Oh, I should check.

File: examples/official-site/sqlpage/migrations/01_documentation.sql (hunk 2 above)
[the remark this answers is not in the document]
Comment (lovasoa) [reply examples/official-site/sqlpage/migrations/01_documentation.sql:426]: That would work. But that would make all forms less efficient for everyone by default. I think your current code change is enough to have something that works and is efficient for everyone by default, yet can be customized.

File: examples/official-site/sqlpage/migrations/01_documentation.sql (hunk 1 above)
Comment (djyotta) [reply examples/official-site/sqlpage/migrations/01_documentation.sql:218]: Sorry, I don't follow. There is only one line here.

Comment (lovasoa) [reply examples/official-site/sqlpage/migrations/01_documentation.sql:218]: I was saying that this line is fine, and we don't need the other changes in the documentation besides it.

Comment (djyotta) [reply examples/official-site/sqlpage/migrations/01_documentation.sql:218]: Please could you make the desired changes? I find the documentation of the row level param and the side-effect of file type input helpful so I'm loathe to remove it.

**Comment by djyotta** [comment]
For reference - here is proposed changes that I think will make the handling of form encoding to work intuitively whilst still allowing flexability for developer to set specific encoding

https://github.com/djyotta/SQLpage/pull/1

File: examples/official-site/sqlpage/migrations/01_documentation.sql (hunk 3 above)
[the remark this answers is not in the document]
Comment (djyotta) [reply examples/official-site/sqlpage/migrations/01_documentation.sql]: Though I was unable to find specification of this behaviour, it seems that anecdotally at least files will not upload without multipart/form-data.

All I found was https://datatracker.ietf.org/doc/html/rfc1867 which is quite old and says that file upload behaviour without multipart/form-data set as encoding type is unspecified: https://datatracker.ietf.org/doc/html/rfc1867#section-5.8

File: examples/official-site/sqlpage/migrations/01_documentation.sql (hunk 2 above)
[the remark this answers is not in the document]
Comment (djyotta) [reply examples/official-site/sqlpage/migrations/01_documentation.sql]: https://github.com/djyotta/SQLpage/pull/1 - here is what I think would be a better way to ensure file uploads work out-of-the-box but without overriding the form enctype or resulting in custom buttons or external buttons lacking the correct formenctype.

[the remark this answers is not in the document]
Comment (djyotta) [reply examples/official-site/sqlpage/migrations/01_documentation.sql]: Resolving as I have removed the reference to potential future changes altogether.

**Review by lovasoa — APPROVED** [review]
Ok, let's merge it like this at the moment, I'll update it later. Thanks for the contribution!

**Comment by djyotta** [comment]
No problem. Thanks for SQLpage making UI easy.

Sorry I couldn't catch the vision on this one. It's hard to see the forrest through the trees sometimes.

**Pull request merged**